### PR TITLE
WIP: migration: merge PDB controller into the Migration Controller

### PR DESCRIPTION
**Pull request by acardace**
As these 2 controllers are strongly coupled let's merge them
to dynamically extend and shrink PDBs (to protect both the source pod
and target pod without races) when doing migrations.

Signed-off-by: Antonio Cardace <acardace@redhat.com>

**What this PR does / why we need it**:
This PR merges the PDB controller into the migration controller so that when starting migrations the PDB protecting a migratable VMI can be enlarged to protect both the source and target pods, and later be shrunk again when the migration has completed, all of this without timing gaps as the migration controller has all the knowledge on when to shrink/extend the PDB.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
This is still a WIP as the legacy code for the PDB controller must still be removed and test adjusted.

**Release note**:

```release-note
PDBs for migratable VMIs are now resized on the fly when migration starts/completes.
```

### CODE REVIEW

File: pkg/virt-controller/watch/migration.go [hunk 1]
```diff
@@ -772,13 +806,347 @@ func (c *MigrationController) listMigrationsMatchingVMI(namespace string, name s
...
+		return
+	}
+
+	curControllerRef := v1.GetControllerOf(curPodDisruptionBudget)
+	oldControllerRef := v1.GetControllerOf(oldPodDisruptionBudget)
+	controllerRefChanged := !reflect.DeepEqual(curControllerRef, oldControllerRef)
+	if controllerRefChanged {
+		// The ControllerRef was changed. Sync the old controller, if any.
+		if vmi := c.resolveControllerVmiRef(oldPodDisruptionBudget.Namespace, oldControllerRef); vmi != nil {
+			if err := c.syncPDB(vmi); err != nil {
```
Comment (davidvossel) [inline pkg/virt-controller/watch/migration.go:930]: we shouldn't be performing any real work in the pod informer handlers.

The reason for this is, if the syncPDB() funciton returns an error, it won't be processed again. In these informer event handler functions, we need to be queueing a key onto the work queue. With the work-queue, if something like syncPDB fails, then the key is re-enqueued and tried again later.

**Review by davidvossel — COMMENTED** [review]
as i'm looking at this in practice, I think this is going to be super tricky.

The migration controller is reconciling migration objects and our pdb controller is reconciling VMI objects. I have no idea how we can successfully merge the two without dealing with a lot of edge conditions. 

for example, one migration can fail, which allows another migration to start, but the pdb contracting and expansion somehow needs to be synced in a way that ensures the pdb is in the right state before the new migration is started. Otherwise if we get the ordering wrong the previous failed migration might attempt to shrink the pdb that was just expanded for a new migration.

Then we have to also account for the fact that a migration's target pod can outlive the migration, which means that we can easily invalidate our PDB if we aren't careful. #5425 aims to solve that though.

**Comment by acardace** [comment]
Going to close this if https://github.com/kubevirt/kubevirt/pull/5460 is considered a better solution.

**Comment by acardace** [comment]
/close

Closing in favor of https://github.com/kubevirt/kubevirt/pull/5460.

**Pull request closed without merging**